src/mumble/JackAudio.cpp

```diff
@@ -1059,6 +1125,92 @@ bool JackAudioOutput::process(const jack_nframes_t frames) {
 	return true;
 }
 
+void JackAudioOutput::prepareOutputBuffers(unsigned int frameCount, QList< AudioOutputBuffer * > &qlMix,
+										   QList< AudioOutputBuffer * > &qlDel) {
+	ServerHandlerPtr sh = Global::get().sh;
+	VoiceRecorderPtr recorder;
+	if (sh) {
+		recorder = Global::get().sh->recorder;
+
+		// Shortcut to base class code when our special case is not needed.
+		if (recorder && !recorder->isTransportEnabled()) {
+			AudioOutput::prepareOutputBuffers(frameCount, qlMix, qlDel);
+			return;
+		}
+	}
+
+	// Register user ports based on who is currently present in qmOutputs and route their audio to JACK.
+	// Don't care if users get removed for now. TODO: maybe at some point we do.
+	auto it = qmOutputs.constBegin();
+	while (it != qmOutputs.constEnd()) {
+		const ClientUser *user   = it.key();
+		AudioOutputBuffer *audio = it.value();
+
+		if (!user || !recorder || (recorder && !recorder->isTransportEnabled())) {
+			if (audio->prepareSampleBuffer(frameCount)) {
+				qlMix.append(audio);
+			} else {
+				qlDel.append(audio);
+			}
+
+			++it;
+			continue;
+		}
+
+		QString qsPortName = user->qsName;
+		qsPortName         = qsPortName.prepend("user_");
+		if (userPorts.find(qsPortName) == userPorts.end()) {
+			jack_port_t *port = jas->registerPort(qsPortName.toStdString().c_str(), JackPortIsOutput);
+			if (!port) {
+				qWarning("JackAudioOutput: unable to register user port \"%s\"", qsPortName.toStdString().c_str());
+			} else {
+				userPorts[qsPortName] = port;
+			}
+		}
+
+		if (userBuffers.find(qsPortName) == userBuffers.end()) {
+			jack_ringbuffer_t *ringBuffer =
+				jas->ringbufferCreate(static_cast< size_t >(iFrameSize) * iSampleSize * JACK_BUFFER_PERIODS);
+			if (!ringBuffer) {
+				qWarning("JackAudioOutput: unable to create user buffer for port \"%s\"",
+						 qsPortName.toStdString().c_str());
+			} else {
+				jas->ringbufferMlock(ringBuffer);
+				userBuffers[qsPortName] = ringBuffer;
+			}
+		}
+
+		auto tempBuffer = std::make_unique< float[] >(frameCount);
```

Comment:
Maybe a static vector would be better here in order to avoid allocating memory in every iteration of this loop. The static vector can be reused across loop iterations and even across function invocations. That means we only allocate if we have to grow that vector

Reply:
Sure, that's easy. Will do